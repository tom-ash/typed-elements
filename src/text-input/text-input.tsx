import React, { ChangeEvent } from "react";

interface TextInputInterface {
  (props: {
    label: string | React.ReactElement;
    id?: string;
    value?: string;
    onChange?(newValue: string, changeEvent?: ChangeEvent): void;
    containerClassName?: string;
    disabled?: boolean;
  }): React.ReactElement;
}

export const TextInput: TextInputInterface = props => {
  const { label, id, value, onChange, containerClassName, disabled } = props;

  return (
    <div className={containerClassName}>
      <label>{label}</label>
      <input
        type="text"
        id={id}
        disabled={disabled}
        value={value}
        onChange={onChange ? e => onChange(e.target.value, e) : undefined}
      />
    </div>
  );
};
