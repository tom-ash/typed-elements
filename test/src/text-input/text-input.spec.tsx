import React from "react";
import { describe, expect, test } from "@jest/globals";
import { TextInput } from "../../../src/text-input/text-input";
import renderer from "react-test-renderer";

describe("Text Input", () => {
  describe("label", () => {
    describe("when the provided label is string", () => {
      test("renders with the provided label", () => {
        const component = renderer.create(<TextInput label="Test" />);

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe("when the provided label is ReactElement", () => {
      test("renders with the provided label", () => {
        const component = renderer.create(<TextInput label={<div>Test</div>} />);

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("controlled input", () => {
    describe("when value and onChange callback are provided", () => {
      test("renders with the provided value and callback", () => {
        const component = renderer.create(
          <TextInput
            label="Test"
            value="test"
            onChange={newValue => console.log(newValue)}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("containerClassName", () => {
    describe("when containerClassName is provided", () => {
      test("renders with the provided containerClassName", () => {
        const component = renderer.create(
          <TextInput
            label="Test"
            value="test"
            onChange={newValue => console.log(newValue)}
            containerClassName="test"
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("id", () => {
    describe("when id is provided", () => {
      test("renders with the provided id", () => {
        const component = renderer.create(
          <TextInput
            id="test"
            label="Test"
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("disabled", () => {
    describe("when disabled is provided", () => {
      test("renders with the provided disabled", () => {
        const component = renderer.create(
          <TextInput
            disabled={true}
            label="Test"
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
