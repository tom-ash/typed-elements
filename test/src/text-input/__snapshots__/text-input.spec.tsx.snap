// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Input containerClassName when containerClassName is provided renders with the provided containerClassName 1`] = `
<div
  className="test"
>
  <label>
    Test
  </label>
  <input
    onChange={[Function]}
    type="text"
    value="test"
  />
</div>
`;

exports[`Text Input controlled input when value and onChange callback are provided renders with the provided value and callback 1`] = `
<div>
  <label>
    Test
  </label>
  <input
    onChange={[Function]}
    type="text"
    value="test"
  />
</div>
`;

exports[`Text Input disabled when disabled is provided renders with the provided disabled 1`] = `
<div>
  <label>
    Test
  </label>
  <input
    disabled={true}
    type="text"
  />
</div>
`;

exports[`Text Input id when id is provided renders with the provided id 1`] = `
<div>
  <label>
    Test
  </label>
  <input
    id="test"
    type="text"
  />
</div>
`;

exports[`Text Input label when the provided label is ReactElement renders with the provided label 1`] = `
<div>
  <label>
    <div>
      Test
    </div>
  </label>
  <input
    type="text"
  />
</div>
`;

exports[`Text Input label when the provided label is string renders with the provided label 1`] = `
<div>
  <label>
    Test
  </label>
  <input
    type="text"
  />
</div>
`;
